ruyi/cli/completion.py

```diff
@@ -0,0 +1,74 @@
+"""
+helper functions for CLI completions
+"""
+
+from typing import cast, Callable, Optional, Protocol, Sequence, Any, TYPE_CHECKING
+import argparse
+
+from argcomplete.completers import BaseCompleter
+
+if TYPE_CHECKING:
+    from .. import config
+
+
+class ArgcompleteAction(argparse.Action):
+    # see https://github.com/kislyuk/argcomplete/issues/443 for why this typing is needed
+    completer: Optional[Callable[[str, object], list[str]]]
+
+    def __call__(
+        self,
+        parser: argparse.ArgumentParser,
+        namespace: argparse.Namespace,
+        values: str | Sequence[Any] | None,
+        option_string: str | None = None,
+    ) -> None:
+        raise NotImplementedError('.__call__() not defined')
+
+
+class SelfArgumentParser(argparse.ArgumentParser):
+    def add_argument(self, *args: Any, **kwargs: Any) -> ArgcompleteAction:
+        return cast(ArgcompleteAction, super().add_argument(*args, **kwargs))
+
+
+class NoneCompleter(BaseCompleter):
+    def __call__(self,
+                 *,
+                 prefix: str,
+                 action: argparse.Action,
+                 parser: argparse.ArgumentParser,
+                 parsed_args: argparse.Namespace) -> None:
+        return None
+
+
+class DynamicCompleter(Protocol):
+    def __call__(self, prefix: str, parsed_args: object,
+                 **kwargs: Any) -> list[str]: ...
+
+
+def package_completer_builder(
+    cfg: "config.GlobalConfig", filters: list[Callable[[str], bool]] | None = None
```

Comment:
In general we put each argument on its own line in this codebase, like this:

```suggestion
    cfg: "config.GlobalConfig",
    filters: list[Callable[[str], bool]] | None = None,
```